# HW1 : Sentiment Analysis on Twitter

### Submission By: Stuti Sanghavi
### Student Id : 00001587699

In [1]:
# print all the outputs in a cell
from IPython.core.interactiveshell import InteractiveShell
InteractiveShell.ast_node_interactivity = "all"

## Part 1: Using the keyword 'Trump'

### Importing packages in Python

In [2]:
import re

### Defining a function to import data

In [3]:
# --------------------------------- Defining a function to import data --------------------------------------------#

def import_data (filename):
    
    #Open a connection with the file
    file = open(filename, mode = 'r')
    
    #Reading the data
    data = file.read()
    
    #Closing the file connection
    file.close()
    
    #Returning the data
    return data

### Importing the Trump Tweet data (Text file) 

In [4]:
# Importing Trump.txt file by calling the function import_data

tweet_data = import_data('Trump.txt')

# Checking part of the data to see if imported correctly
tweet_data[0:500]

'b"RT @TheRickWilson: 1/ I\'ll summarize this for the slow learners in the Senate one last time. \\n\\nYour protestations of independence, integrit\\xe2\\x80\\xa6\\n"b\'RT @alexespind: Madame Speaker is correct. There\\xe2\\x80\\x99s a #GopCoverup . Let\\xe2\\x80\\x99s get it trending. https://t.co/zChKIc6GzP\\n\'b"RT @RepAdamSchiff: Trump\'s lawyer claimed the House isn\'t ready to present our case.\\n\\nWe\\xe2\\x80\\x99re ready.\\n\\nThe House calls John Bolton to testify.\\xe2\\x80\\xa6\\n"b\'RT @Jamierodr14: GOOD! \\n\\nRI'

### Cleaning the data

#### a. Removing the @mentions in the tweet data

In [5]:
#Removing @mentions

tweet_data_2 = re.sub(r'@[A-Za-z0-9]+','',tweet_data)

#### b. Removing url links from the data

In [6]:
#Removing url links

tweet_data_3 = re.sub('https?://[A-Za-z0-9./]+','',tweet_data_2)

#### c. Removing new line, breaks, retweets etc.

In [7]:
#Replacing new line, breaks, retweets with blank spaces

tweet_data_4 = tweet_data_3.replace('\n',' ').replace('\'b',' ').replace('\'RT',' ').replace('\\n', ' ').replace('"RT',' ').replace('"b',' ').replace('https:/',' ').replace('_','')

#### d. Removing emoticons (eg. \\xe2\\x80)

In [8]:
#Removing emoticons

tweet_data_5 = re.sub(r'\\[a-z0-9A-Z]+','',tweet_data_4)

#### e. Removing hashtags, punctuations, etc.

In [9]:
#Getting rid of hashtags and punctuations

tweet_data_6 = re.sub(r'[^\w\s]','',tweet_data_5)

#### f. Removing numbers from the data

In [10]:
#Removing numbers from data

tweet_data_7 = re.sub(r'[0-9]+','',tweet_data_6)

#### g. Converting all the data to lowercase for getting the final cleaned file

In [11]:
# Converting data to lower case 
tweet_data_cleaned = tweet_data_7.lower()

#Displaying first 300 characters of the cleaned data
tweet_data_cleaned[0:300]

'b    ill summarize this for the slow learners in the senate one last time   your protestations of independence integrit     madame speaker is correct there a gopcoverup  let get it trending      trumps lawyer claimed the house isnt ready to present our case  we ready  the house calls john bolton to '

### Getting the clean data in form of a list

In [51]:
#Getting the cleaned tweet data stored in a list form
tweet_list = tweet_data_cleaned.strip().split()

# Displaying only 10 words of the list
tweet_list[5:15]

#Displaying cleaned word count
len(tweet_list)

['the',
 'slow',
 'learners',
 'in',
 'the',
 'senate',
 'one',
 'last',
 'time',
 'your']

236233

### Importing positive, negative and stop words list

#### a. Importing positive words list by calling the function import_data

In [13]:
positive_words = import_data('positive.txt')

#### b. Converting the text file to list and displaying the results

In [14]:
#Converting the text file(positive_words) to a list
positive_list = positive_words.strip().split()

#Displaying only the first 10 words from the positive list
positive_list[0:10]


['a+',
 'abound',
 'abounds',
 'abundance',
 'abundant',
 'accessable',
 'accessible',
 'acclaim',
 'acclaimed',
 'acclamation']

#### c. Importing negative words list

In [15]:
negative_words = import_data('negative.txt')

#### d. Converting the text file to list and displaying the results

In [16]:
#Converting the text file(negative_words) to a list
negative_list = negative_words.strip().split()

# Displaying only the first 10 words from the negative list
negative_list[0:10]

['2-faced',
 '2-faces',
 'abnormal',
 'abolish',
 'abominable',
 'abominably',
 'abominate',
 'abomination',
 'abort',
 'aborted']

#### e. Importing stop words list

In [17]:
stop_words = import_data('stop.txt')

#### f. Converting the text file to list and displaying the results

In [18]:
#Converting the text file(stop_words) to a list
stop_list = stop_words.strip().split()

#Displaying only the first 10 words from the stop list
stop_list[0:10]

['a',
 'about',
 'above',
 'across',
 'after',
 'afterwards',
 'again',
 'against',
 'all',
 'almost']

### Running a loop for each word

I defined a function called sentiments, which takes input of tweet_data, positive words, neagtive words and stop words.
The counters for positive, negative, stop and others have been set to 0. After which, a 'for' loop is run, in which, for a word in the tweet_data if the word is found in the positive, then the counter for positive words will increase by 1. Similarly, the counters for negative, stop and others would increase if the word from the tweet data is found in any of the word list. This will continue till all the words from the tweet_list are counted. Once that is done, the loop returns the count for each word type

### Defining a function called sentiments for counting, positive, negative, stop and other words

In [19]:
#Defining a function called sentiments for categorizing and counting, positive, negative, stop and other words

def sentiments(tweet_list, positive_list, negative_list, stop_list):
    count_positive = 0
    count_negative = 0
    count_stop = 0
    count_others = 0

    for word in tweet_list:
        if word in positive_list:
            count_positive += 1
        elif word in negative_list:
            count_negative += 1
        elif word in stop_list:
            count_stop += 1
        else:
            count_others += 1
    
    all_counts = [count_positive,count_negative, count_stop, count_others]
    return all_counts

In [20]:
# Calling the sentiments function on tweet data

total_counts = sentiments(tweet_list, positive_list, negative_list, stop_list)
total_counts

[14728, 9170, 104523, 107812]

### Finding the ratio and percentage of positive/total word count

In [21]:
print("Ratio of positive to total word count is " + str(total_counts[0]/len(tweet_list)))

Ratio of positive to total word count is 0.06234522695813032


In [22]:
print("Percentage of positive word count is " + str(total_counts[0]/len(tweet_list)*100))

Percentage of positive word count is 6.234522695813032


### Finding the ratio and percentage of negative/total word count

In [23]:
print("Ratio of negative to total word count is " + str(total_counts[1]/len(tweet_list)))

Ratio of negative to total word count is 0.03881760803952031


In [24]:
print("Percentage of negative word count is " + str(total_counts[1]/len(tweet_list)*100))

Percentage of negative word count is 3.8817608039520306


### Finding the ratio and percentage of stop/total word count

In [25]:
print("Ratio of stop to total word count is " + str(total_counts[2]/len(tweet_list)))

Ratio of stop to total word count is 0.4424572350179696


In [26]:
print("Percentage of stop word count is " + str(total_counts[2]/len(tweet_list)*100))

Percentage of stop word count is 44.24572350179696


### Finding the ratio and percentage of other/total word count

In [27]:
print("Ratio of other to total word count is " + str(total_counts[3]/len(tweet_list)))

Ratio of other to total word count is 0.4563799299843798


In [28]:
print("Percentage of other word count is " + str(total_counts[3]/len(tweet_list)*100))

Percentage of other word count is 45.63799299843798


### Defining a function called sentiment output for interpreting the sentiment

I defined a function called sentiment_output, which takes input of positive count and negative count. Then there is an if else statement which is used to categorize and analyze the sentiment attached to every word. 
- If the count_positive >= 1.75 that of negative then it is a very strong positive sentiment and it will print the same
- Similarly, there are other conditions used to categorize sentiment as mildly positive, neutral, strongly negative and mildly negative.
- If one of the condition in the if statement is not satisfied, it will go on to the next, until one of the conditions is satisfied and it throws an ouput.

In [29]:
#Defining a function called sentiment_output for interpreting the sentiment and categorizing it

def sentiment_output(count_positive, count_negative):
    if count_positive >= 1.75 * count_negative:
        print("The sentiment is Strongly positive")
    elif count_positive >= 1.10 * count_negative:
        print("The sentiment is Mildly positive")
    elif count_negative >= 1.75 * count_positive:
        print("The sentiment is Strongly negative")
    elif count_negative >= 1.10 * count_positive:
        print("The sentiment is Mildly negative")
    else:
        print("The sentiment is Neutral")

### Calling the sentiment_output function below to see the nature of the sentiment

In [30]:
#Inputting the positive and negative word counts
sentiment_output(total_counts[0], total_counts[1])

The sentiment is Mildly positive


### Verifying the result from above, by taking the ratio of total positive to total negative word count

In [31]:
output = total_counts[0]/total_counts[1]
round(output,2)

1.61

We get the ratio as 1.61, which falls under the range of mildly positive (as defined above) and hence, the result is verified

### Converting list to tuple and running %timeit on it to check which one takes lesser time

In [32]:
#-----------------------------------Using tuples instead of lists------------------------------------------#

tweet_tuple = tuple(tweet_list)
positive_tuple = tuple(positive_list)
negative_tuple = tuple(negative_list)
stop_tuple = tuple(stop_list)


In [50]:
#--------------------------- Running %timeit on it to check which runs faster---------------------------------------------#

%timeit sentiments(tweet_tuple, positive_tuple, negative_tuple, stop_tuple)
%timeit sentiments(tweet_list, positive_list, negative_list, stop_list)

34.4 s ± 691 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)
34.1 s ± 964 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)


Running %timeit on tuple and list, and comparing the results

## Conclusion: The overall sentiment is mildly positive for the keyword Trump

## Part 2 : For the keyword "Schiff"

### Importing the Tweet data (Text file) by using the import_data function

In [34]:
# Importing Schiff.txt file by calling the function import_data

tweet_data_schiff = import_data('Schiff.txt')

# Checking part of the data to see if imported correctly
tweet_data_schiff[0:500]

'b\'RT @Trump2Usa: "Ridiculous!" - Trump Lead Impeachment Lawyer Pat Cipollone DESTROYS Schiff (VIDEO) https://t.co/nx1qpYLxhW\\n\'b\'RT @RubenGallego: The Vice President\\xe2\\x80\\x99s brother had access to the SCIF and Intel. He sits on one of the three Committees that were allowed i\\xe2\\x80\\xa6\\n\'b\'RT @JRubinBlogger: In his calm and methodical presentation, Schiff scored a victory: Democrats have effectively built the case that the Sen\\xe2\\x80\\xa6\\n\'b"RT @RepAndyBiggsAZ: Schiff Argument #1: We can\'t'

### Cleaning the data

In [35]:
# ------------------------------------------ Cleaning the data ----------------------------------------------------

# Removing @mentions in the data
tweet_data1 = re.sub(r'@[A-Za-z0-9]+','',tweet_data_schiff)

#Removing url links from the data
tweet_data2 = re.sub('https?://[A-Za-z0-9./]+','',tweet_data1)

#Removing the new line, breaks, retweets etc.
tweet_data3 = tweet_data2.replace('\n',' ').replace('\'b',' ').replace('\'RT',' ').replace('\\n', ' ').replace('"RT',' ').replace('"b',' ').replace('https:/',' ').replace('_','')

# Removing emoticons (eg. \xe2\x80)
tweet_data4 = re.sub(r'\\[a-z0-9A-Z]+','',tweet_data3)

#Removing hashtags, punctuations, etc.
tweet_data5 = re.sub(r'[^\w\s]','',tweet_data4)

#Removing numbers from the data
tweet_data6 = re.sub(r'[0-9]+','',tweet_data5)

#Converting all the data to lowercase for getting the final cleaned file
schiff_data_cleaned = tweet_data6.lower()

#Displaying first 500 characters of the cleaned data
schiff_data_cleaned[0:500]

'b   ridiculous  trump lead impeachment lawyer pat cipollone destroys schiff video      the vice president brother had access to the scif and intel he sits on one of the three committees that were allowed i     in his calm and methodical presentation schiff scored a victory democrats have effectively built the case that the sen     schiff argument  we cant have a fair senate trial wo the documents amp witnesses we didnt call in the house when we  schiff and all the dem aholes hate our president s'

### Getting the clean data in form of a list

In [52]:
#Converting the cleaned data into a list of words by splitting the data
schiff_data_list = schiff_data_cleaned.strip().split()

# Displaying only 10 words of the list
schiff_data_list[5:15]

#Displaying cleaned word count
len(schiff_data_list)

['lawyer',
 'pat',
 'cipollone',
 'destroys',
 'schiff',
 'video',
 'the',
 'vice',
 'president',
 'brother']

241943

### Running a loop for each word

Here we will use the same list of positive, negative and stop words and run a for loop on the cleaned data using those lists. For doing that, we do not need to code the same thing again, but merely calling the sentiment function with the schiff_data_list as the input along with the other lists

In [37]:
# Calling the sentiments function on schiff_data_list

total_counts_schiff = sentiments(schiff_data_list, positive_list, negative_list, stop_list)
total_counts_schiff

[8075, 11165, 105037, 117666]

### Finding the ratio and percentage of positive/total word count

In [38]:
print("Ratio of positive to total word count is " + str(total_counts_schiff[0]/len(schiff_data_list)))

Ratio of positive to total word count is 0.03337562979710097


In [39]:
print("Percentage of positive word count is " + str(total_counts_schiff[0]/len(schiff_data_list)*100))

Percentage of positive word count is 3.337562979710097


### Finding the ratio and percentage of negative/total word count

In [40]:
print("Ratio of negative to total word count is " + str(total_counts_schiff[1]/len(schiff_data_list)))

Ratio of negative to total word count is 0.04614723302596066


In [41]:
print("Percentage of negative word count is " + str(total_counts_schiff[1]/len(schiff_data_list)*100))

Percentage of negative word count is 4.614723302596066


### Finding the ratio and percentage of stop/total word count

In [42]:
print("Ratio of stop to total word count is " + str(total_counts_schiff[2]/len(schiff_data_list)))

Ratio of stop to total word count is 0.43413944606787547


In [43]:
print("Percentage of stop word count is " + str(total_counts_schiff[2]/len(schiff_data_list)*100))

Percentage of stop word count is 43.413944606787545


### Finding the ratio and percentage of other/total word count

In [44]:
print("Ratio of other to total word count is " + str(total_counts_schiff[3]/len(schiff_data_list)))

Ratio of other to total word count is 0.4863376911090629


In [45]:
print("Percentage of other word count is " + str(total_counts_schiff[3]/len(schiff_data_list)*100))

Percentage of other word count is 48.633769110906286


### Calling the sentiment_output function from above to see the nature of the sentiment

In [46]:
#Inputting the positive and negative word counts
sentiment_output(total_counts_schiff[0], total_counts_schiff[1])

The sentiment is Mildly negative


### Verifying the result from above, by taking the ratio of total negative to total positive word count

In [47]:
output = total_counts_schiff[1]/total_counts_schiff[0]
round(output,2)

1.38

*We get the ratio as 1.38, which falls under the range of mildly negative (as defined in the function above) and hence, the result is verified

### Converting list to tuple and running %timeit on it to check which one takes lesser time

In [48]:
#-----------------------------------Using tuples instead of lists------------------------------------------#

schiff_data_tuple = tuple(schiff_data_list)

# We do not need to convert positive, negative or stop word list to tuple since they are already converted in part 1 
# and hence we can use the same tuple 

In [49]:
#--------------------------- Running %timeit on it to check which runs faster---------------------------------------------#

%timeit sentiments(schiff_data_tuple, positive_tuple, negative_tuple, stop_tuple)
%timeit sentiments(schiff_data_list, positive_list, negative_list, stop_list)

35.2 s ± 2.33 s per loop (mean ± std. dev. of 7 runs, 1 loop each)
35.2 s ± 1.98 s per loop (mean ± std. dev. of 7 runs, 1 loop each)


Running %timeit on tuple and list, and comparing the results

## Conclusion: The overall sentiment is mildly negative for the keyword Schiff